lib/api/objectPutCopyPart.js

```diff
@@ -0,0 +1,356 @@
+import async from 'async';
+import { errors } from 'arsenal';
+
+import constants from '../../constants';
+import data from '../data/wrapper';
+import kms from '../kms/wrapper';
+import metadata from '../metadata/wrapper';
+import RelayMD5Sum from '../utilities/RelayMD5Sum';
+import { logger } from '../utilities/logger';
+import routesUtils from '../routes/routesUtils';
+import services from '../services';
+import { parseRange } from './apiUtils/object/parseRange';
+import validateHeaders from '../utilities/validateHeaders';
+
+
+/**
+ * PUT Part Copy during a multipart upload.
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with
+ * requester's info
+ * @param {request} request - request object given by router,
+ *                            includes normalized headers
+ * @param {string} sourceBucket - name of source bucket for object copy
+ * @param {string} sourceObject - name of source object for object copy
+ * @param {object} log - the log request
+ * @param {function} callback - final callback to call with the result
+ * @return {undefined}
+ */
+export default
+function objectPutCopyPart(authInfo, request, sourceBucket,
+    sourceObject, log, callback) {
+    log.debug('processing request', { method: 'objectPutCopyPart' });
+    const destBucketName = request.bucketName;
+    const destObjectKey = request.objectKey;
+    const valGetParams = {
+        authInfo,
+        bucketName: sourceBucket,
+        objectKey: sourceObject,
+        requestType: 'objectGet',
+        log,
+    };
+
+
+    const partNumber = Number.parseInt(request.query.partNumber, 10);
+    // AWS caps partNumbers at 10,000
+    if (partNumber > 10000 || !Number.isInteger(partNumber) || partNumber < 1) {
+        return callback(errors.InvalidArgument);
+    }
+    // We pad the partNumbers so that the parts will be sorted
+    // in numerical order
+    const paddedPartNumber = `000000${partNumber}`.substr(-5);
+    // Note that keys in the query object retain their case, so
+    // request.query.uploadId must be called with that exact
+    // capitalization
+    const uploadId = request.query.uploadId;
+
+    const valPutParams = {
+        authInfo,
+        bucketName: destBucketName,
+        objectKey: destObjectKey,
+        requestType: 'objectPut',
+        log,
+    };
+
+    // For validating the request at the MPU, the params are the same
+    // as validating for the destination bucket except additionally need
+    // the uploadId and splitter.
+    // Also, requestType is 'putPart or complete'
+    const valMPUParams = Object.assign({
+        uploadId,
+        splitter: constants.splitter,
+    }, valPutParams);
+    valMPUParams.requestType = 'putPart or complete';
```

Comment:
I know that this is being used by complete MPU as well, we could address this in a different PR but we should make this all upper-cased or all lower-cased, otherwise it's easy to mess up the case.